### ISSUE TITLE
Comparing cftime.datetime(0, 0, 0, calendar=None)

### ISSUE THREAD

**Issue by bjlittle** [Unidata/cftime#248]
## Question
For a `cftime.datetime` where the `calendar=None`, we have a corner case where the user is able to create a valid `datetime` without explicitly specifying `has_year_zero` e.g.,
```python
>>> cftime.datetime(0, 0, 0, calendar=None)
cftime.datetime(0, 0, 0, 0, 0, 0, 0, calendar='', has_year_zero=False)
```
However, this returns a `cftime.datetime` object with a `has_year_zero=False`. Although this is correct given what the user originally specified, the `year` of the `cftime.datetime` is clearly **not** zero.

As a result, this causes issues when comparing such a `cftime.datetime` to one created with `has_year_zero=True` (see the following **Example** section).

Would it be possible to ensure that the `has_year_zero` property is verified and correct given the `year` at creation time and also if the user alters the `year` though the `replace` method i.e., keep the `has_year_zero` internally consistent given the `year` value of the instance? 

## Example
Existing behaviour:
```python
from cftime import datetime

>>> a = datetime(0, 0, 0, calendar=None)
>>> b = datetime(0, 0, 0, calendar=None, has_year_zero=True)
>>> a == a
True
>>> b == b
True
>>> a == b
TypeError: cannot compare cftime.datetime(0, 0, 0, 0, 0, 0, 0, calendar='', has_year_zero=False) and cftime.datetime(0, 0, 0, 0, 0, 0, 0, calendar='', has_year_zero=True)
```
```python
>>> dt = datetime(1970, 1, 1, calendar=None)
>>> dt
cftime.datetime(1970, 1, 1, 0, 0, 0, 0, calendar='', has_year_zero=False)
>>> dt.replace(year=0)
cftime.datetime(0, 1, 1, 0, 0, 0, 0, calendar='', has_year_zero=False)
```
Proposed behaviour:
```python
from cftime import datetime

>>> a = datetime(0, 0, 0, calendar=None)
>>> b = datetime(0, 0, 0, calendar=None, has_year_zero=True)
>>> a == a
True
>>> b == b
True
>>> a == b
True
```
```python
>>> dt = datetime(1970, 1, 1, calendar=None)
>>> dt
cftime.datetime(1970, 1, 1, 0, 0, 0, 0, calendar='', has_year_zero=False)
>>> dt.replace(year=0)
cftime.datetime(0, 1, 1, 0, 0, 0, 0, calendar='', has_year_zero=True)
```

## Context
- Version: `cftime` 1.5.0
- Platform: `linux`

**Comment by jswhit** [comment]
Validating the year at creation time should not be a problem.  The replace method creates a new instance, so it would then trigger an error if you tried to replace the year with 0 and has_year_zero is False.  Or are you suggesting that has_year_zero would be reset to True in that case?

**Comment by bjlittle** [comment]
@jswhit Thanks for taking the time to look at this :+1: 

My initial instinct would be that in the new instance created by `replace` the `year=0` is what the user wants, and `has_year_zero` is reset appropriately i.e., `has_year_zero=True`. The `has_year_zero` is derived from the provided value of `year` and set accordingly so that the `cftime.datetime` instance is internally consistent and correct. 

Clearly, if the user explicitly does:
- `cftime.datetime(0, 0, 0, calendar=None, has_year_zero=False)`
- `cftime.datetime(0, 1, 1, has_year_zero=False)`
- `cftime.datetime(1, 0, 0, calendar=None, has_year_zero=True)`
- `cftime.datetime(1, 1, 1, has_year_zero=True)`

Then an exception is raised. But:
- `cftime.datetime(0, 0, 0, calendar=None)` will result in an instance with `has_year_zero=True` automatically set
- `cftime.datetime(1, 0, 0, calendar=None)` will result in an instance with `has_year_zero=False` automatically set

I guess this is achievable under-the-hood by giving the `has_year_zero` formal parameter a default value of `None`, that way you know if the user has set it (`True`|`False`), and if not calculate the correct value based on the `year` provided (and the choice of `calendar` may factor into this, I'm not sure, but I expect you know that :wink:)

I guess I'm suggesting that it seem entirely possible (and convenient for the user) for `cftime` to derive the correct value of `has_year_zero` based on the provided `year` when `has_year_zero` **hasn't** been explicitly specified.

Does that help clarify?

Does this behaviour seem reasonable and something that should be adopted?

**Comment by jswhit** [comment]
WIP at PR #248.  Needs tests.

**Comment by jswhit** [comment]
I am a little concerned that with this update the `has_year_zero` attribute could change silently, if for example some operation is performed that changes the year to zero when `has_year_zero=False`.  I don't think this can happen, but there might be a corner case I'm missing.

**Comment by jswhit** [comment]
PR #249 is ready to be merged, pending a review by @bjlittle

**Comment by jswhit** [comment]
Closed by PR #249
